hubble/cmd/common/conn/conn.go

```diff
@@ -30,13 +34,100 @@ var GRPCOptionFuncs []GRPCOptionFunc
 func init() {
 	GRPCOptionFuncs = append(
 		GRPCOptionFuncs,
-		grpcInterceptors,
+		grpcUnaryInterceptors,
+		grpcStreamInterceptors,
 		grpcOptionTLS,
 	)
 }
 
-func grpcInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
-	return grpc.WithUnaryInterceptor(timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout))), nil
+func grpcUnaryInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainUnaryInterceptor(
+		timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout)),
+		logVersionMismatchUnaryInterceptor(),
+	)
+	return option, nil
+}
+
+func grpcStreamInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainStreamInterceptor(
+		logVersionMismatchStreamInterceptor(),
+	)
+	return option, nil
+}
+
+func logVersionMismatchUnaryInterceptor() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req any, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		var header metadata.MD
+		opts = append(opts, grpc.Header(&header))
+		if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+			return err
+		}
+		logVersionMismatch(header)
+		return nil
+	}
+}
+
+func logVersionMismatchStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		stream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		// stream.Header() blocks until metadata is ready to read.
+		// This could be forever if no metadata is ever sent, after
+		// the first read from the stream or after an explicit call
+		// to SendHeader() server-side.
+		go func() {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			header, err := stream.Header()
+			if err == nil {
+				logVersionMismatch(header)
+			}
+		}()
+		return stream, nil
+	}
+}
+
+func logVersionMismatch(header metadata.MD) {
+	var relayVersion string
+	relayVersionHeaders := header.Get(relaydefaults.GRPCMetadataRelayVersionKey)
+	if len(relayVersionHeaders) > 0 {
+		relayVersion = relayVersionHeaders[0]
+	}
+	var serverVersion string
+	serverVersionHeaders := header.Get(serverdefaults.GRPCMetadataServerVersionKey)
+	if len(serverVersionHeaders) > 0 {
+		serverVersion = serverVersionHeaders[0]
+	}
+
+	relayVersionMissing := relayVersion == ""
+	relayVersionMismatch := relayVersion != build.RelayVersion.SemVer()
+	serverVersionMissing := serverVersion == ""
+	serverVersionMismatch := serverVersion != build.ServerVersion.SemVer()
```

Comment:
I think we should do a semver check here and only log when

```
hubble-cli-version < hubble-{server,relay}-version
```
i.e. using a more recent version of the CLI than the Relay/server version should not log anything as it should be fully supported. We should definitely ignore semver pre-release and build metadata parts (.e.g `-dev` suffix) and I even wonder whether we should ignore the patch level part. What do you think?

Reply:
Yeah I agree. Just ended a discussion with Chance about that and alert fatigue would be strong for this one.

Reply:
Updated to match the conditional: `hubble-cli-version < hubble-{server,relay}-version` and ignore lower parts than Minor.